packages/material-ui-styles/src/styled/styled.d.ts

```diff
@@ -12,14 +12,16 @@ import * as React from 'react';
 export type ComponentCreator<Component extends React.ElementType> = <Theme, Props extends {} = any>(
   styles:
     | CreateCSSProperties<Props>
-    | ((props: { theme: Theme } & CoerceEmptyInterface<Props>) => CreateCSSProperties<Props>),
+    | (({ theme, ...props }: { theme: Theme } & Props) => CreateCSSProperties<Props>),
   options?: WithStylesOptions<Theme>,
 ) => React.ComponentType<
   Omit<
     JSX.LibraryManagedAttributes<Component, React.ComponentProps<Component>>,
     'classes' | 'className'
   > &
-    StyledComponentProps<'root'> & { className?: string } & CoerceEmptyInterface<Props>
+    StyledComponentProps<'root'> & { className?: string } & CoerceEmptyInterface<
+      Props extends { theme: Theme } ? Omit<Props, 'theme'> & { theme?: Theme } : Props
```

Comment:
I don't quite get why this is needed, if the user provides props where theme is required, why do we change it to optional?

Reply:
@merceyz https://codesandbox.io/s/stoic-yonath-3ukmk

Reply:
I will replace the commit when we resolve this, added tests also for core/styles/styled

Reply:
Yeah, I saw the type error but my point still stands.
The types you want to change should be `@material-ui/system`s `style` function, as it is the one saying theme is required.

Reply:
@merceyz 
Quick answer:
A: https://codesandbox.io/s/stoic-yonath-3ukmk
B: https://codesandbox.io/s/crazy-forest-7m5di
For A: typing complains, runtime works because theme is satisfied. You see right theme color resolved. 
For B: typing complains, runtime throws warning (see console), color not resolved.

Long answer:
Produced `StyleFunction` has no direct relation to React/Component. Its just a callable that takes some args and gives some result. It just happens to be that its arg is inferred as prop requirement in Components enhanced using the given `StyleFunction`. The function itself does require theme. It will never get the theme any other way. Contrary a Component get receive theme as prop via context, therefore a `styled` component does not always need theme passed as prop (even thought it always need theme).
`StyleFunction` has no relation to component. Its simply a function which has arg requirement and gives back some result. Its just happens to be that its args are inferred as prop requirement of components which are enhanced using this `StyleFunction`. In this relation `StyleFunction` is consumed without ever knowing about the consumer

Reply:
There are two different props here: The props from the component and the props required in the style function. The that are required in the resulting component are the required props from the component + the required props from the style function without! the theme. Because the theme for the style function will be injected.